### PR TITLE
feat: get supported timezones

### PR DESCRIPTION
**Pull request by ZEED2468**
### Implement Endpoint to Get and Create  Timezones

**Description:**
This PR introduces new endpoints in the `TimezonesController` for creating and retrieving timezones. The endpoints include:
1. An endpoint to create a new timezone.
2. An endpoint to retrieve the list of supported timezones.

**Endpoints:**

**1. Create Timezone**
- **Method:** POST
- **URL:** `/api/v1/timezones`
- **Description:** Creates a new timezone entry in the system.

**Request Body:**
```json
{
  "timezone": "string",
  "gmtOffset": "string",
  "description": "string"
}
```

**Response:**

*Success Response:*
- **Status Code:** 201
- **Body:**
  ```json
  {
    "status_code": 201,
    "message": "Time Created Successfully",
    "timezone": {
        "timezone": "string",
        "gmtOffset": "string",
        "description": "string",
        "id": "uuid"
    }
  }
  ```

*Error Response:*
- **Status Code:** 409
- **Body:**
  ```json
  {
    "status_code": 409,
    "message": "Timezone already exists"
  }
  ```

**2. Get Supported Timezones**
- **Method:** GET
- **URL:** `/api/v1/timezones`
- **Description:** Retrieves a list of all timezones supported by the system.

**Headers:**
- `Authorization: Bearer <token>`

**Response:**

*Success Response:*
- **Status Code:** 200
- **Body:**
  ```json
  {
    "status_code": 200,
    "message": "Timezone fetched successfully",
    "timezones": [
      {
        "id": "uuid",
        "created_at": "string",
        "updated_at": "string",
        "timezone": "string",
        "gmtOffset": "string",
        "description": "string"
      }
    ]
  }
  ```

*Error Response (Unauthorized):*
- **Status Code:** 401
- **Body:**
  ```json
  {
    "status_code": 401,
    "message": "Unauthorized"
  }
  ```

**How to Test Manually:**

1. **Create Timezone:**
   - **Request Type:** POST
   - **Endpoint:** `/api/v1/timezones`
   - **Request Body:** Include the `timezone`, `gmtOffset`, and `description` fields.
   - **Success Response:** Should return the created timezone with a status code of 201.
   - **Error Response:** If the timezone already exists, should return a 409 status code with an appropriate error message.

2. **Fetch Timezones:**
   - **Request Type:** GET
   - **Endpoint:** `/api/v1/timezones`
   - **Headers:** Include a valid `Authorization` token in the request headers.
   - **Success Response:** Should return a list of supported timezones with a status code of 200.
   - **Error Response:** If unauthorized or with an invalid token, should return a 401 status code with the appropriate error message.

**Checklist:**
- [x] Implemented the POST `/api/v1/timezones` endpoint to create a new timezone.
- [x] Implemented the GET `/api/v1/timezones` endpoint to retrieve the list of supported timezones.
- [x] Updated the `TimezonesService` to handle both creating and retrieving timezones.
- [x] Created unit tests for both endpoints, including:
  - [x] Testing successful creation and retrieval of timezones.
  - [x] Handling errors for duplicate timezones and unauthorized access.
- [x] Verified that the endpoints return the correct data format and status codes.
- [x] Added error handling for unauthorized access and existing timezones.
- [x] Updated API documentation with new endpoint details and authentication requirements.

### Reference Issue

[FEAT]: Support for Language Selection [34](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/34)

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -31,10 +85,9 @@ async function bootstrap() {
   app.enableCors();
   app.setGlobalPrefix('api/v1', { exclude: ['/', 'health', 'api', 'api/v1', 'api/docs'] });
 
-  // TODO: set options for swagger docs
   const options = new DocumentBuilder()
-    .setTitle('<project-title-here>')
-    .setDescription('<project-description-here>')
+    .setTitle('Swagger')
```
Comment (CFCIfe) [inline src/main.ts:89]: This is not meant to be. The project Title has been set in a different PR

File: src/main.ts [hunk 2]
```diff
@@ -31,10 +85,9 @@ async function bootstrap() {
...
+    .setDescription('Timezones')
```
Comment (CFCIfe) [inline src/main.ts:90]: Same with this. Please revert

File: src/main.ts [hunk 3]
```diff
@@ -1,3 +1,57 @@
...
+//   const port = app.get<ConfigService>(ConfigService).get<number>('server.port');
+//   await app.listen(port);
+
+//   logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api/v1` });
+// }
+// bootstrap().catch(err => {
+//   console.error('Error during bootstrap', err);
+//   process.exit(1);
+// });
+
```
Comment (CFCIfe) [inline src/main.ts:54]: You need to revert this

File: src/main.ts [hunk 4]
```diff
@@ -45,8 +98,9 @@ async function bootstrap() {
   const port = app.get<ConfigService>(ConfigService).get<number>('server.port');
   await app.listen(port);
 
-  logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api/v1` });
+  logger.log({ message: 'Server started 🚀', port, url: `http://localhost:${port}/api/v1` });
```
Comment (CFCIfe) [inline src/main.ts:101]: You should not be touching shared files that do not affect your implementation

File: src/modules/timezones/timezones.controller.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    description: 'Unauthorized',
+    schema: {
+      example: {
+        status: 'error',
+        message: 'Unauthorized',
+        statusCode: 401,
+      },
+    },
+  })
+  async getTimezonesWithoutAuth(): Promise<{
```
Comment (CFCIfe) [inline src/modules/timezones/timezones.controller.ts:43]: If this is a public route you need to add skipAuth() from the helpers folder.

File: src/modules/timezones/entities/timezone.entity.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
+
+@Entity()
+export class Timezone {
```
Comment (CFCIfe) [inline src/modules/timezones/entities/timezone.entity.ts:4]: Is this a standalone schema? Could not find it here - https://app.eraser.io/workspace/d1jC73I9ARAXST9QM2do

File: src/modules/timezones/timezones.service.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+          status_code: HttpStatus.INTERNAL_SERVER_ERROR,
+        },
+        HttpStatus.INTERNAL_SERVER_ERROR
+      );
+    }
+  }
+
+  async getSupportedTimezones(): Promise<any> {
+    try {
+      const timezones = await this.timezoneRepository.find();
```
Comment (CFCIfe) [inline src/modules/timezones/timezones.service.ts:57]: This does not look like it would work.

File: src/main.ts (hunk 1 above)
Comment (ZEED2468) [reply src/main.ts:89]: okay, I will do that now

File: src/main.ts (hunk 2 above)
Comment (ZEED2468) [reply src/main.ts:90]: okay

File: src/main.ts (hunk 3 above)
Comment (ZEED2468) [reply src/main.ts:54]: okay

File: src/main.ts (hunk 4 above)
Comment (ZEED2468) [reply src/main.ts:101]: okay will do work on it

File: src/modules/timezones/entities/timezone.entity.ts (hunk 6 above)
Comment (ZEED2468) [reply src/modules/timezones/entities/timezone.entity.ts:4]: yeah, it's a standalone schema

**Comment by ZEED2468** [comment]
completed everything

File: src/modules/timezones/dto/create-timezone.dto.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { IsString, IsOptional } from 'class-validator';
+
+export class CreateTimezoneDto {
```
Comment (Cobidex) [inline src/modules/timezones/dto/create-timezone.dto.ts:3]: Add descriptive swagger docs to the dto fields

File: src/modules/timezones/entities/timezone.entity.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
```
Comment (Cobidex) [inline src/modules/timezones/entities/timezone.entity.ts:4]: All classes that require id and metadata(updated_at and created_at) field should extend the base class to keep codebase DRY.

File: src/modules/timezones/timezones.controller.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { Controller, Get, Post, Body, Res, HttpCode, HttpStatus } from '@nestjs/common';
+import { TimezonesService } from './timezones.service';
+import { CreateTimezoneDto } from './dto/create-timezone.dto';
+import { Response } from 'express';
+
+@Controller('timezones')
```
Comment (Cobidex) [inline src/modules/timezones/timezones.controller.ts:6]: Add docs to your controller

File: src/modules/timezones/timezones.controller.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+export class TimezonesController {
+  constructor(private readonly timezonesService: TimezonesService) {}
+
+  @Post()
+  async createTimezone(@Body() createTimezoneDto: CreateTimezoneDto, @Res() response: Response) {
```
Comment (Cobidex) [inline src/modules/timezones/timezones.controller.ts:11]: Add swagger docs

File: src/modules/timezones/timezones.controller.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    const createdTimezone = await this.timezonesService.createTimezone(createTimezoneDto);
+    return response.status(createdTimezone.status_code).send(createdTimezone);
+  }
+
+  @Get()
+  async getTimezones(@Res() response: Response): Promise<any> {
```
Comment (Cobidex) [inline src/modules/timezones/timezones.controller.ts:17]: Add docs

Comment (ZEED2468) [reply src/modules/timezones/timezones.controller.ts:26]: done

File: src/modules/timezones/timezones.controller.ts (hunk 11 above)
Comment (ZEED2468) [reply src/modules/timezones/timezones.controller.ts:17]: done

File: src/modules/timezones/timezones.controller.ts (hunk 10 above)
Comment (ZEED2468) [reply src/modules/timezones/timezones.controller.ts:8]: done

File: src/modules/timezones/entities/timezone.entity.ts (hunk 9 above)
Comment (ZEED2468) [reply src/modules/timezones/entities/timezone.entity.ts]: done

**Review by TheCodeGhinux — DISMISSED** [review]
Fair, fix conflicts

**Review by Cobidex — APPROVED** [review]

File: src/modules/timezones/entities/timezone.entity.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
+import { AbstractBaseEntity } from '../../../entities/base.entity'; // Assume you have a BaseEntity defined
+
+@Entity()
+export class Timezone extends AbstractBaseEntity {
+  @PrimaryGeneratedColumn('uuid')
```
Comment (Homoakin619) [inline src/modules/timezones/entities/timezone.entity.ts:6]: Extending the AbstractBaseEntity provides this already, kindly remove the id property definition

Comment (ZEED2468) [reply src/modules/timezones/entities/timezone.entity.ts]: Done

File: src/main.ts [hunk 14]
```diff
@@ -33,7 +33,6 @@ async function bootstrap() {
   app.setGlobalPrefix('api/v1', { exclude: ['/', 'health', 'api', 'api/v1', 'api/docs'] });
   app.useGlobalInterceptors(new ResponseInterceptor());
 
-  // TODO: set options for swagger docs
```
Comment (CFCIfe) [inline src/main.ts:36]: Revert this. If yo remove it you need to DO it

File: src/modules/timezones/entities/timezone.entity.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
+import { AbstractBaseEntity } from '../../../entities/base.entity'; // Assume you have a BaseEntity defined
```
Comment (CFCIfe) [inline src/modules/timezones/entities/timezone.entity.ts:2]: remove comment

File: src/modules/timezones/timezones.service.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    try {
+      const timezoneExists = await this.timezoneRepository.findOne({
+        where: { timezone: createTimezoneDto.timezone },
+      });
+
+      if (timezoneExists) {
+        return {
+          status_code: HttpStatus.CONFLICT,
+          message: TIMEZONE_ALREADY_EXISTS,
+        };
```
Comment (CFCIfe) [inline src/modules/timezones/timezones.service.ts:31]:
```suggestion
throw new ConflictException({
        status_code: HttpStatus.CONFLICT,
        message: TIMEZONE_ALREADY_EXISTS,
    })
```

File: src/modules/timezones/timezones.service.ts [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+      };
+    } catch (error) {
+      Logger.error('TimezonesServiceError ~ createTimezone ~', error);
+      throw new HttpException(
+        {
+          message: ERROR_OCCURED,
+          status_code: HttpStatus.INTERNAL_SERVER_ERROR,
+        },
+        HttpStatus.INTERNAL_SERVER_ERROR
+      );
```
Comment (CFCIfe) [inline src/modules/timezones/timezones.service.ts:50]:
```suggestion
throw new InternalServerErrorException({
          message: ERROR_OCCURED,
          status_code: HttpStatus.INTERNAL_SERVER_ERROR,
        });
```

File: src/modules/timezones/timezones.service.ts [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+      };
+    } catch (error) {
+      Logger.error('TimezonesServiceError ~ fetchTimezones ~', error);
+      throw new HttpException(
+        {
+          message: FETCH_TIMEZONE_FAILURE,
+          status_code: HttpStatus.INTERNAL_SERVER_ERROR,
+        },
+        HttpStatus.INTERNAL_SERVER_ERROR
+      );
```
Comment (CFCIfe) [inline src/modules/timezones/timezones.service.ts:70]:
```suggestion
throw new InternalServerErrorException({
          message: FETCH_TIMEZONE_FAILURE,
          status_code: HttpStatus.INTERNAL_SERVER_ERROR,,
        });
```

**Review by CFCIfe — COMMENTED** [review]
Resolve the comments when done

**Review by TheCodeGhinux — APPROVED** [review]

**Review by TheCodeGhinux — APPROVED** [review]

**Review by Cobidex — APPROVED** [review]

**Review by Homoakin619 — APPROVED** [review]

**Pull request merged**